### ISSUE TITLE
Pyright does not understand when `==` and `!=` get used with a primary data type as the first check

### ISSUE THREAD

**Issue by mmohaveri** [microsoft/pyright#3622]
**Describe the bug**
If you overload `==` and `!=` by default pyright respects the type hint of the overloaded function's return type. But if you use it to compare a primitive with that type and put the primitive as the first argument, pyright assumes the output is boolean.

**To Reproduce**
```pyright
class Data(object):
    def __eq__(self, other: object, /) -> str:
        return "string"

    def __ne__(self, other: "Data| bool", /) -> str:
        return "string"


B = Data()

b1 = (B == False)
b2 = (B == B)
b3 = (False == B)

reveal_type(b1)  # str
reveal_type(b2)  # str
reveal_type(b3)  # bool

b4 = (B != False)
b5 = (B != B)
b6 = (False != B)

reveal_type(b4)  # str
reveal_type(b5)  # str
reveal_type(b6)  # bool
```
**Expected behavior**
As `__eq__` does not have a `__req__` companion and python automatically uses the correct overload I expect pyright to respect the return type even when the primitive type comes first.

**Screenshots or Code**
![3](https://user-images.githubusercontent.com/6017468/175933768-efa93289-0786-4775-9810-aa684652e5a5.png)

**VS Code extension or command-line**
Issue happens in both CLI and VSCode extenstion. pyright version 1.1.255.

**Comment by erictraut** [comment]
Pyright is doing the right thing here, at least with the type information it is given.

When evaluating the type of an `==` operator, pyright first determines the types of the left-hand-side (LHS) and RHS operands. It then determines whether a compatible `__eq__` method is implemented by the LHS type. If not, it then looks for a compatible `__ne__` method on the RHS type. This is the same process used at runtime by the Python interpreter.

Now, let's consider the expression `False == B`. The LHS type is `bool`, and the RHS type is `Data`. The `bool` class inherits its `__eq__` method from the `int` class, which is defined in the [typeshed stub](https://github.com/python/typeshed/blob/5b44454cf84405dcdbb2ddc0a9b7ca5f71ffa689/stdlib/builtins.pyi#L289) `builtins.pyi` as follows:

```python
    def __eq__(self, __x: object) -> bool: ...
```

Since the RHS type is compatible with `object`, this `__eq__` method is assumed to be a match, and it returns a `bool`.

At runtime, if I evaluate `False.__eq__(Data)`, the method returns `NotImplemented`. That tells me that the `int.__eq__` method doesn't actually accept any `object` but rather accepts only `int` or a subclass thereof. So this could be considered a bug in the typeshed definition of `int.__eq__`. You could consider filing a bug in the [typeshed issue tracker](https://github.com/python/typeshed/issues). However, I suspect that the typeshed maintainers will say that the current definition of `int.__eq__` is intended, since it is not uncommon for a caller to pass an `object` argument.

In summary, I think pyright is doing the right thing here with the type information it is give.

What you're trying to do here — to override the `__eq__` method to return a `str` is quite out of the ordinary, so it's not surprising to me that you're running into limitations like this. Perhaps you could find a different way to accomplish what you're doing here?

**Comment by mmohaveri** [comment]
Thanks, I'll report the issue in typeshed issue tracker. IMO type hint is suppose to inform users (and type checkers) about correct ways to call a function. If a function returns a `NotImplemented` for an input type, that type should not be included in its type hint (unless its impossible to type hint the input, for example in a case that `NotImplemented` rises from a specific combination of input types, not a single input per se.

P.S: I'm not trying to return an `str`, that's just an example. What I'm trying to do it to create a function like `__eq__` in numpy's `NDArray`. There, if you try to call `__eq__` on an `NDArray`, you''ll get back an `NDArray` since equality check is being done in an element-wise manner. I'm trying to replicate that.
